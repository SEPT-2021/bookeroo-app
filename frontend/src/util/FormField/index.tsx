import React from "react";
import { TextField, TextFieldProps } from "@material-ui/core";

/**
 * A wrapper around TextField with easy input to a map of errors, which automatically enters an errored state if an error is present
 */
function FormField({ onChange, name, errors, ...rest }: FormFieldProps) {
  return (
    <TextField
      {...rest}
      name={name}
      error={!!errors?.[name]}
      helperText={errors?.[name]}
      variant="outlined"
      margin="normal"
      required
      fullWidth
      onChange={(e) => onChange && onChange(e.target.value)}
    />
  );
}

type FormFieldProps<T extends string = string> = Omit<
  TextFieldProps,
  "onChange"
> & {
  onChange?: (val: string) => void;
  name: T;
  // Not strict types, but we do expect name to be a possible key of errors and data
  errors?: Record<string | T, string>;
};

export default FormField;
