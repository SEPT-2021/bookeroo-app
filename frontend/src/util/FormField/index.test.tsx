import React from "react";
import { render } from "@testing-library/react";
import FormField from ".";

test("renders normal", () => {
  const { container } = render(<FormField name="test" />);
  expect(container).toMatchSnapshot();
});
test("renders with other error", () => {
  const { container } = render(
    <FormField name="test" errors={{ other: "No Error" }} />
  );
  expect(container).toMatchSnapshot();
});
test("renders with real error", () => {
  const { container } = render(
    <FormField name="test" errors={{ test: "actual error" }} />
  );
  expect(container).toMatchSnapshot();
});
