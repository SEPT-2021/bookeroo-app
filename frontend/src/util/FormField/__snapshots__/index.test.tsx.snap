// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders normal 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input"
        name="test"
        required=""
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend-2"
          style="width: 0.01px;"
        >
          <span>
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`renders with other error 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input"
        name="test"
        required=""
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend-6"
          style="width: 0.01px;"
        >
          <span>
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`renders with real error 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <input
        aria-invalid="true"
        class="MuiInputBase-input MuiOutlinedInput-input"
        name="test"
        required=""
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend-10"
          style="width: 0.01px;"
        >
          <span>
            ​
          </span>
        </legend>
      </fieldset>
    </div>
    <p
      class="MuiFormHelperText-root MuiFormHelperText-contained Mui-error Mui-required"
    >
      actual error
    </p>
  </div>
</div>
`;
